core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+    fn calculate_my_repairman_index_for_slot(
+        my_id: &Pubkey,
+        eligible_repairmen: &[&Pubkey],
+        num_blobs_in_slot: usize,
+        repair_redundancy: usize,
+    ) -> Option<(usize, usize)> {
+        let total_blobs = num_blobs_in_slot * repair_redundancy;
+        let total_repairmen = min(total_blobs, eligible_repairmen.len());
+
+        let blobs_per_repairman = min(
+            (total_blobs + total_repairmen - 1) / total_repairmen,
+            num_blobs_in_slot,
+        );
+
+        // Calculate the indexes this node is responsible for
+        if let Some(my_position) = eligible_repairmen[..total_repairmen]
+            .iter()
+            .position(|id| *id == my_id)
+        {
+            let begin = (my_position * blobs_per_repairman) % num_blobs_in_slot;
+            Some((begin, blobs_per_repairman))
+        } else {
+            // If there are more repairmen than `total_blobs`, then some repairmen
+            // will not have any responsibility to repair this slot
+            None
+        }
+    }
+
+    fn find_eligible_repairmen<'a>(
+        repairee_root: u64,
+        repairman_roots: &'a HashMap<Pubkey, (u64, u64)>,
+        epoch_schedule: &EpochSchedule,
+        num_buffer_slots: usize,
+    ) -> Vec<&'a Pubkey> {
+        let repairmen: Vec<_> = repairman_roots
+            .iter()
+            .filter_map(|(repairman_id, (_, repairman_root))| {
+                if Self::should_repair_peer(
+                    *repairman_root,
+                    repairee_root,
+                    epoch_schedule,
+                    num_buffer_slots,
+                ) {
+                    Some(repairman_id)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        repairmen
+    }
+
+    fn update_my_gossiped_root(
+        my_id: &Pubkey,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        old_root: &mut u64,
+    ) -> u64 {
+        let new_root = cluster_info
+            .read()
+            .unwrap()
+            .get_gossiped_root_for_node(&my_id, None);
+
+        if let Some(new_root) = new_root {
+            *old_root = new_root;
+            new_root
+        } else {
+            *old_root
+        }
+    }
+
+    // Decide if a repairman with root == `repairman_root` should send repairs to a
+    // potential repairee with root == `repairee_root`
+    fn should_repair_peer(
+        repairman_root: u64,
+        repairee_root: u64,
+        epoch_schedule: &EpochSchedule,
+        num_buffer_slots: usize,
+    ) -> bool {
+        // Check if this potential repairman's confirmed leader schedule is greater
+        // than an epoch ahead of the repairee's known schedule
+        let repairman_epoch = epoch_schedule.get_stakers_epoch(repairman_root);
+        let repairee_epoch =
+            epoch_schedule.get_stakers_epoch(repairee_root + num_buffer_slots as u64);
```

Comment:
can you explain this bit of code?  what does num_buffer_slots mean?

Reply:
The way repairmen decide if a repairee qualifies for repair is by checking whether `get_stakers_epoch(my_root) > get_stakers_epoch(repairee_root + num_buffer_slots)`. This is to prevent repairing repairees that are only slightly behind. For instance if:

1) Epochs are 32 slots long
2) A repairman is at root 16
3) A repairee is at root 15

Then `get_stakers_epoch` for the repairman will be 1, and the `get_stakers_epoch` for the repairee will be 0, but a repair is probably not necessary b/c the repairee is not sufficiently far behind.

Reply:
num_buffer_slots should come from epoch schedule, then?

Reply:
It's specifically for the repair use case though?

Reply:
yeah.  this is a little messy...  maybe let's skip the epoch check and merely run on roots

under the above proposal sentence ;):  the repairman is happy to throw the next slot at the repairee as long as the repairee is asking for a slot that's in the repairman's root hierarchy

Reply:
so throw out the check for whether you should repair a particular repairee based on epoch, and instead just repair any node on the network that is missing a slot in your root hierarchy?

Reply:
yeah...